02_Util/src/security/SignedMeterValuesUtil.ts

```diff
@@ -0,0 +1,209 @@
+import {
+  IChargingStationSecurityInfoRepository,
+  IDeviceModelRepository,
+  sequelize,
+} from '@citrineos/data';
+import {
+  IFileAccess,
+  MeterValueType,
+  SignedMeterValuesConfig,
+  SignedMeterValueType,
+  SystemConfig,
+} from '@citrineos/base';
+import { ILogObj, Logger } from 'tslog';
+import * as crypto from 'node:crypto';
+
+/**
+ * Util to process and validate signed meter values.
+ */
+export class SignedMeterValuesUtil {
+  private readonly _fileAccess: IFileAccess;
+  private readonly _logger: Logger<ILogObj>;
+  private readonly _deviceModelRepository: IDeviceModelRepository;
+  private readonly _chargingStationSecurityInfoRepository: IChargingStationSecurityInfoRepository;
+
+  private readonly _signedMeterValuesConfiguration:
+    | SignedMeterValuesConfig
+    | undefined;
+
+  /**
+   * @param {IFileAccess} [fileAccess] - The `fileAccess` allows access to the configured file storage.
+   *
+   * @param {SystemConfig} config - The `config` contains the current system configuration settings.
+   *
+   * @param {Logger<ILogObj>} [logger] - The `logger` represents an instance of {@link Logger<ILogObj>}.
+   *
+   * @param {IDeviceModelRepository} [deviceModelRepository] - The `deviceModelRepository` represents a repository for accessing and manipulating variable attribute data.
+   *
+   */
+  constructor(
+    fileAccess: IFileAccess,
+    config: SystemConfig,
+    logger: Logger<ILogObj>,
+    deviceModelRepository: IDeviceModelRepository,
+  ) {
+    this._fileAccess = fileAccess;
+    this._logger = logger;
+    this._deviceModelRepository = deviceModelRepository;
+    this._chargingStationSecurityInfoRepository =
+      new sequelize.SequelizeChargingStationSecurityInfoRepository(
+        config,
+        logger,
+      );
+
+    this._signedMeterValuesConfiguration =
+      config.modules.transactions.signedMeterValuesConfiguration;
+  }
+
+  public async validateMeterValues(
+    stationId: string,
+    meterValues: [MeterValueType, ...MeterValueType[]],
+  ): Promise<boolean> {
+    let validMeterValues = true;
+
+    const expectPublicKey =
+      await this._deviceModelRepository.readAllByQuerystring({
+        stationId,
+        component_name: 'OCPPCommCtrlr',
```

Comment:
since these are part of the spec, would it make sense to create some enums around the component name and the variable names? I see that these are the possible values (below), so maybe it may be helpful to have as enum since they are not truly dynamic strings... 
```
ControllerEnum {
  AlignedDataCtrlr
  AuthCacheCtrlr
  AuthCtrlr
  CHAdeMOCtrlr
  ClockCtrlr
  CustomizationCtrlr
  DeviceDataCtrlr
  DisplayMessageCtrlr
  ISO15118Ctrlr
  LocalAuthListCtrlr
  MonitoringCtrlr
  OCPPCommCtrlr
  ReservationCtrlr
  SampledDataCtrlr
  SecurityCtrlr
  SmartChargingCtrlr
  TariffCostCtrlr
  TxCtrlr
}

VariableName list
//-2.1-General
2.1.1.-ActiveNetworkProfile
2.1.2.-AllowNewSessionsPendingFirmwareUpdate
2.1.3.-DefaultMessageTimeout
2.1.4.-FileTransferProtocols
2.1.5.-HeartbeatInterval
2.1.6.-NetworkConfigurationPriority
2.1.7.-NetworkProfileConnectionAttempts
2.1.8.-OfflineThreshold
2.1.9.-QueueAllMessages
2.1.10.-MessageAttemptsTransactionEvent
2.1.11.-MessageAttemptIntervalTransactionEvent
2.1.12.-UnlockOnEVSideDisconnect
2.1.13.-WebSocketPingInterval
2.1.14.-ResetRetries
2.1.15.-MessageFieldLength
2.1.16.-ItemsPerMessageGetReport
2.1.17.-ItemsPerMessageGetVariables
2.1.18.-BytesPerMessageGetReport
2.1.19.-BytesPerMessageGetVariables
2.1.20.-ConfigurationValueSize
2.1.21.-ReportingValueSize
2.1.22.-ItemsPerMessageSetVariables
2.1.23.-BytesPerMessageSetVariables
2.1.24.-DateTime
2.1.25.-NtpSource
2.1.26.-NtpServerUri
2.1.27.-TimeOffset
2.1.28.-NextTimeOffsetTransitionDateTime
2.1.29.-TimeOffsetNextTransition
2.1.30.-TimeSource
2.1.31.-TimeZone
2.1.32.-TimeAdjustmentReportingThreshold
2.1.33.-CustomImplementationEnabled
//-2.2-Security-related
2.2.1.-BasicAuthPassword
2.2.2.-Identity
2.2.3.-OrganizationName
2.2.4.-CertificateEntries
2.2.5.-SecurityProfile
2.2.6.-AdditionalRootCertificateCheck
2.2.7.-MaxCertificateChainSize
2.2.8.-CertSigningWaitMinimum
2.2.9.-CertSigningRepeatTimes
//-2.3-Authorization-related
2.3.1.-AuthEnabled
2.3.2.-AdditionalInfoItemsPerMessage
2.3.3.-OfflineTxForUnknownIdEnabled
2.3.4.-AuthorizeRemoteStart
2.3.5.-LocalAuthorizeOffline
2.3.6.-LocalPreAuthorize
2.3.7.-MasterPassGroupId
2.3.8.-DisableRemoteAuthorization
//-2.4-Authorization-Cache-related
2.4.1.-AuthCacheEnabled
2.4.2.-AuthCacheAvailable
2.4.3.-AuthCacheLifeTime
2.4.4.-AuthCacheStorage
2.4.5.-AuthCachePolicy
2.4.6.-AuthCacheDisablePostAuthorize
//-2.5-Local-Authorization-List-Management-related
2.5.1.-LocalAuthListEnabled
2.5.2.-LocalAuthListEntries
2.5.3.-LocalAuthListAvailable
2.5.4.-ItemsPerMessageSendLocalList
2.5.5.-BytesPerMessageSendLocalList
2.5.6.-LocalAuthListStorage
2.5.7.-LocalAuthListDisablePostAuthorize
2.5.8.-LocalAuthListSupportsExpiryDateTime
//-2.6-Transaction-related
2.6.1.-EVConnectionTimeOut
2.6.2.-StopTxOnEVSideDisconnect
2.6.3.-TxBeforeAcceptedEnabled
2.6.4.-TxStartPoint
2.6.5.-TxStopPoint
2.6.6.-TxStartStopPoint-values
2.6.6.1.-TxStartPoint-values
2.6.6.2.-TxStopPoint-values
2.6.7.-MaxEnergyOnInvalidId
2.6.8.-StopTxOnInvalidId
//-2.7-Metering-related
2.7.1.-SampledDataEnabled
2.7.2.-SampledDataAvailable
2.7.3.-SampledDataSignReadings
2.7.4.-SampledDataTxEndedMeasurands
2.7.5.-SampledDataTxEndedInterval
2.7.6.-SampledDataTxStartedMeasurands
2.7.7.-SampledDataTxUpdatedMeasurands
2.7.8.-SampledDataTxUpdatedInterval
2.7.9.-AlignedDataEnabled
2.7.10.-AlignedDataAvailable
2.7.11.-AlignedDataMeasurands
2.7.12.-AlignedDataInterval
2.7.13.-AlignedDataSendDuringIdle
2.7.14.-AlignedDataSignReadings
2.7.15.-AlignedDataTxEndedMeasurands
2.7.16.-AlignedDataTxEndedInterval
2.7.17.-PublicKeyWithSignedMeterValue
2.7.18.-SampledDataRegisterValuesWithoutPhases
//-2.8-Reservation-related
2.8.1.-ReservationEnabled
2.8.2.-ReservationAvailable
2.8.3.-ReservationNonEvseSpecific
//-2.9-Smart-Charging-related
2.9.1.-SmartChargingEnabled
2.9.2.-SmartChargingAvailable
2.9.3.-ACPhaseSwitchingSupported
2.9.4.-ChargingProfileMaxStackLevel
2.9.5.-ChargingScheduleChargingRateUnit
2.9.6.-PeriodsPerSchedule
2.9.7.-ExternalControlSignalsEnabled
2.9.8.-NotifyChargingLimitWithSchedules
2.9.9.-Phases3to1
2.9.10.-ChargingProfileEntries
2.9.11.-LimitChangeSignificance
//-2.10-Tariff-&-Cost-related
2.10.1.-TariffEnabled
2.10.2.-TariffAvailable
2.10.3.-TariffFallbackMessage
2.10.4.-CostEnabled
2.10.5.-CostAvailable
2.10.6.-TotalCostFallbackMessage
2.10.7.-Currency
//-2.11-Diagnostics-related
2.11.1.-MonitoringEnabled
2.11.2.-MonitoringAvailable
2.11.3.-ItemsPerMessageClearVariableMonitoring
2.11.4.-ItemsPerMessageSetVariableMonitoring
2.11.5.-BytesPerMessageClearVariableMonitoring
2.11.6.-BytesPerMessageSetVariableMonitoring
2.11.7.-OfflineMonitoringEventQueuingSeverity
2.11.8.-ActiveMonitoringBase
2.11.9.-ActiveMonitoringLevel
//-2.12-Display-Message-related
2.12.1.-DisplayMessageEnabled
2.12.2.-DisplayMessageAvailable
2.12.3.-NumberOfDisplayMessages
2.12.4.-DisplayMessageSupportedFormats
2.12.5.-DisplayMessageSupportedPriorities
//-2.13-Charging-Infrastructure-related
2.13.1.-Available
2.13.2.-AvailabilityState
2.13.3.-AllowReset
2.13.4.-ConnectorType
2.13.5.-PhaseRotation
2.13.6.-SupplyPhases
2.13.7.-Power
2.13.8.-Example-Reporting-of-EVSEs-and-Connectors-via-device-model
//-2.14-ISO-15118-Related
2.14.1.-CentralContractValidationAllowed
2.14.2.-ContractValidationOffline
2.14.3.-ProtocolSupportedByEV
2.14.4.-ProtocolAgreed
2.14.5.-ISO15118PnCEnabled
2.14.6.-ISO15118V2GCertificateInstallationEnabled
2.14.7.-ISO15118ContractCertificateInstallationEnabled
2.14.8.-ISO15118RequestMeteringReceipt
2.14.9.-ISO15118SeccId
2.14.10.-ISO15118CountryName
2.14.11.-ISO15118OrganizationName
2.14.12.-ISO15118EvseId
```

Reply:
oooh yes i wondered this as well! i'll look into making a task and creating a separate PR for this... 🤔

Reply:
I'm fine with adding enums as references, please do not type any objects or methods with these enums--the component and variable names should always be strings. This is important to maintain the flexibility that the device model operates on as a core principle. 

The enums will make it a lot easier to access/recall the known components/variables 👍

Reply:
a task has been created, will resolve this thread, thanks!

Reply:
In this case, if it is possible for these strings to further extend the possible values defined in the enum, then should we create a type that extends both enum and string like:
```
export type Components = ComponentsEnum | string;
```
using a type like this would allow for the code to be able to support both enums and other string values...